src/main/drivers/sonar_srf10.c

```diff
@@ -0,0 +1,160 @@
+
+#define SRF10_COMMAND_InitiateRangingInches 0x50
+#define SRF10_COMMAND_InitiateRangingCm 0x51
+#define SRF10_COMMAND_InitiateRangingMicroSeconds 0x53
+
+#define SRF10_COMMAND_SetGain_40 0x00
+#define SRF10_COMMAND_SetGain_100 0x06
+#define SRF10_COMMAND_SetGain_200 0x09
+#define SRF10_COMMAND_SetGain_300 0x0B
+#define SRF10_COMMAND_SetGain_400 0x0D
+#define SRF10_COMMAND_SetGain_500 0x0E
+#define SRF10_COMMAND_SetGain_600 0x0F
+#define SRF10_COMMAND_SetGain_700 0x10 // default and maximum
+#define SRF10_COMMAND_SetGain_Max 0x10 // default
+
+#define SRF10_COMMAND_ChangeAddress1 0xA0
+#define SRF10_COMMAND_ChangeAddress2 0xAA
+#define SRF10_COMMAND_ChangeAddress3 0xA5
+
+// The range is (RangeRegister + 1) * 43mm
+#define SRF10_RangeValue43mm 0
+#define SRF10_RangeValue86mm 1
+#define SRF10_RangeValue1m 24
+#define SRF10_RangeValue4m 93
+#define SRF10_RangeValue6m 139 // maximum range
+#define SRF10_RangeValue11m 0xFF // exceeds actual maximum range
+
+STATIC_UNIT_TESTED volatile int32_t srf10measurementCm = SONAR_OUT_OF_RANGE;
+static int16_t minimumFiringIntervalMs;
+static uint32_t timeOfLastMeasurementMs;
+
+#ifdef UNIT_TEST
+bool i2cWrite(uint8_t addr_, uint8_t reg, uint8_t data) {UNUSED(addr_); UNUSED(reg); UNUSED(data); return false;}
+bool i2cRead(uint8_t addr_, uint8_t reg, uint8_t len, uint8_t* buf) {UNUSED(addr_); UNUSED(reg);UNUSED(len); UNUSED(buf); return false;}
+#endif
+
+static bool i2c_srf10_send_command(uint8_t command)
+{
+    return i2cWrite(SRF10_AddressI2C, SRF10_WRITE_CommandRegister, command);
+}
+
+static bool i2c_srf10_send_byte(uint8_t i2cRegister, uint8_t val)
+{
+    return i2cWrite(SRF10_AddressI2C, i2cRegister, val);
+}
+
+static uint8_t i2c_srf10_read_byte(uint8_t i2cRegister)
+{
+    uint8_t byte;
+    i2cRead(SRF10_AddressI2C, i2cRegister, 1, &byte);
+    return byte;
+}
+
+bool srf10_detect()
+{
+    const uint8_t value = i2c_srf10_read_byte(SRF10_READ_Unused);
+    if (value == SRF10_READ_Unused_ReturnValue) {
+        return true;
+    }
+    return false;
+}
+
+void srf10_init(sonarRange_t *sonarRange)
+{
+    sonarRange->maxRangeCm = SRF10_MAX_RANGE_CM;
+    sonarRange->detectionConeDeciDegrees = SRF10_DETECTION_CONE_DECIDEGREES;
+    sonarRange->detectionConeExtendedDeciDegrees = SRF10_DETECTION_CONE_EXTENDED_DECIDEGREES;
+    // set up the SRF10 hardware for a range of 6m
+    minimumFiringIntervalMs = SRF10_MinimumFiringIntervalFor600cmRangeMs;
+    i2c_srf10_send_byte(SRF10_WRITE_MaxGainRegister, SRF10_COMMAND_SetGain_Max);
+    i2c_srf10_send_byte(SRF10_WRITE_RangeRegister, SRF10_RangeValue6m);
+    // initiate first ranging command
+    i2c_srf10_send_command(SRF10_COMMAND_InitiateRangingCm);
+    timeOfLastMeasurementMs = millis();
+}
+
+/*
+ * Start a range reading
+ * Called periodically from within main process loop
+ */
+void srf10_start_reading(void)
+{
+    // check if there is a measurement outstanding, 0xFF is returned if no measurement
```

Comment:
0xff is returned as long as measurement is pending.

Sending another command when measurement is pending has no effect (srf10 i2c is disabled)

```
const uint8_t revision = i2c_srf10_read_byte(SRF10_READ_SoftwareRevision);
if (revision == 0xFF)
   return; // measurement in progress
```

The ACK part may cause problem if I2C driver is changed/improved in future and no ACK on device addressing is  handled differently